### PR TITLE
Supported hardware serial, update RegisterID.ino for example

### PR DESCRIPTION
**Pull request by HMKRL**
Tested on Arduino MEGA 2560 && Arduino UNO R3 with example RegisterID

### CODE REVIEW

**Comment by LanKuDot** [comment]
Consider using `dynamic_cast` to check the child class.
Check this example: [CheckChildType.cpp](https://gist.github.com/LanKuDot/9bcea6d21978fef65ed63d951b9eca48)
Reference: http://stackoverflow.com/questions/7190952/access-child-members-within-parent-class-c

**Comment by HMKRL** [comment]
Arduino IDE has disabled RTTI (`-fno-rtti`) so `dynamic_cast` or `typeid` won't work.

File: BRCClient/examples/RegisterID/RegisterID.ino [hunk 1]
```diff
@@ -20,7 +20,11 @@
 #define TCP_PORT   5000
 #define MY_COMM_ID 0x20
 
-BRCClient brcClient(UART_RX, UART_TX);
+// if you're using software serial, uncomment this line.
+// BRCClient brcClient(UART_RX, UART_TX);
+
+// if you're using hardware serial, uncomment this line.
+// BRCClient brcClient(&Serial3); //use "TX3" and "RX3" on arduino mega
```
Comment (LanKuDot) [inline BRCClient/examples/RegisterID/RegisterID.ino:27]: Try to use marco to select which constructor to be used.

File: KSM111_ESP8266/KSM111_ESP8266.cpp [hunk 2]
```diff
@@ -11,6 +11,18 @@
...
+	_resetPin = resetPin;
+	_serial = new SoftwareSerial(rxPin, txPin);
+}
+
+KSM111_ESP8266::KSM111_ESP8266(HardwareSerial *hws, int resetPin) {
+	_serialType = HARD;
+	_resetPin = resetPin;
+	_serial = hws;
+}
+
```
Comment (LanKuDot) [inline KSM111_ESP8266/KSM111_ESP8266.cpp:25]: Follow the coding style. The definition of the constructors is in the header file, and try to use initialization list to initialize the private members.

File: BRCClient/BRCClient.h [hunk 3]
```diff
@@ -2,6 +2,7 @@
 #define _BRC_CLIENT_H_
 
 #include <KSM111_ESP8266.h>
+#include <HardwareSerial.h>
```
Comment (LanKuDot) [inline BRCClient/BRCClient.h:5]: `HardwareSerial.h` has been already included in the `KSM111_ESP8266.h`.

**Comment by LanKuDot** [comment]
Hey, I left some comments on your change log. And I pushed some commits after this pull request submitted. Could you modify the code from the latest version? Thank you.

**Comment by HMKRL** [comment]
Ok, I'll make change on latest version.
I'll add macro for #define MEGA_HARD and #define MEGA_SOFT

**Comment by LanKuDot** [comment]
I thought using `USE_HARDWARE_SERIAL` is better.

**Comment by HMKRL** [comment]
Understand, and should I use `SERIAL(X)` to let user choose which serial
port they want to use?

**Comment by LanKuDot** [comment]
I didn't understand. Could you give me some example?

**Pull request closed without merging**